app/workers/messaging.php

```diff
@@ -20,59 +30,180 @@
-            'vonage' => new Vonage($user, $secret),
+        $credentials = $record->getAttribute('credentials');
+        return match ($record->getAttribute('provider')) {
+            'mailgun' => new Mailgun($credentials['apiKey'], $credentials['domain'], $credentials['isEuRegion']),
+            'sendgrid' => new SendGrid($credentials['apiKey']),
             default => null
         };
+    }
 
-        $this->from = App::getEnv('_APP_SMS_FROM');
+    public function init(): void
+    {
     }
 
     public function run(): void
     {
-        if (empty(App::getEnv('_APP_SMS_PROVIDER'))) {
-            Console::info('Skipped sms processing. No Phone provider has been set.');
-            return;
+        $project = new Document($this->args['project']);
+        $this->dbForProject = $this->getProjectDB($project);
+
+        $messageRecord = $this->dbForProject->getDocument('messages', $this->args['messageId']);
+
+        $providerId = $messageRecord->getAttribute('providerId');
+        $providerRecord = $this->dbForProject->getDocument('providers', $providerId);
+
+        $this->processMessage($messageRecord, $providerRecord);
+    }
+
+    private function processMessage(Document $messageRecord, Document $providerRecord): void
+    {
+        $provider = match ($providerRecord->getAttribute('type')) {
+            'sms' => $this->sms($providerRecord),
+            'push' => $this->push($providerRecord),
+            'email' => $this->email($providerRecord),
+            default => null
+        };
+
+        $recipientsId = $messageRecord->getAttribute('to');
+
+        /**
+        * @var Document[] $recipients
+        */
+        $recipients = [];
+
+
+        $topics = $this->dbForProject->find('topics', [Query::equal('$id', $recipientsId)]);
+        foreach ($topics as $topic) {
+            $recipients = \array_merge($recipients, $topic->getAttribute('targets'));
         }
 
-        if (empty($this->from)) {
-            Console::info('Skipped sms processing. No phone number has been set.');
-            return;
+        $users = $this->dbForProject->find('users', [Query::equal('$id', $recipientsId)]);
+        foreach ($users as $user) {
+            $recipients = \array_merge($recipients, $user->getAttribute('targets'));
         }
 
-        $message = new SMS(
-            to: [$this->args['recipient']],
-            content: $this->args['message'],
-            from: $this->from,
-        );
+        $targets = $this->dbForProject->find('targets', [Query::equal('$id', $recipientsId)]);
+        $recipients = \array_merge($recipients, $targets);
 
-        try {
-            $this->sms->send($message);
-        } catch (\Exception $error) {
-            throw new Exception('Error sending message: ' . $error->getMessage(), 500);
+        $identifiers = \array_map(function (Document $recipient) {
+            return $recipient->getAttribute('identifier');
+        }, $recipients);
```

Comment:
We need to filter out all the targets that don't belong to the right provider before we do this

Reply:
filtering the targets now